salt/cloud/clouds/xen.py

```diff
@@ -0,0 +1,1091 @@
+        session = _get_session()
+    log.info('Starting VM {0}'.format(name))
+    vm = _get_vm(name,session)
+    task = session.xenapi.Async.VM.start(vm, False, True)
+    while session.xenapi.task.get_status(task) == 'pending':
+        progress = session.xenapi.task.get_progress(task)
+        log.debug('Waiting for VM to start')
+        time.sleep(2)
+    return show_instance(name)
+
+def pause(name,call=None,session=None):
+    '''
+    Pause a vm 
+
+    .. code-block:: bash
+
+        salt-cloud -a pause xenvm01
+
+    '''
+    if call == 'function':
+        raise SaltCloudException(
+                'The show_instnce function must be called with -a or --action.'
+        )
+    if session is None:
+        session = _get_session()
+    log.info('Pausing VM {0}'.format(name))
+    vm = _get_vm(name,session)
+    pause_vm_task = session.xenapi.Async.VM.pause(vm)
+    while session.xenapi.task.get_status(pause_vm_task) == 'pending':
+        progress = session.xenapi.task.get_progress(pause_vm_task)
+        time.sleep(1)
+    return show_instance(name)
+
+def unpause(name,call=None,session=None):
+    '''
+    UnPause a vm 
+
+    .. code-block:: bash
+
+        salt-cloud -a unpause xenvm01
+
+    '''
+    if call == 'function':
+        raise SaltCloudException(
+                'The show_instnce function must be called with -a or --action.'
+        )
+    if session is None:
+        session = _get_session()
+    log.info('Unpausing VM {0}'.format(name))
+    vm = _get_vm(name,session)
+    task = session.xenapi.Async.VM.unpause(vm)
+    while session.xenapi.task.get_status(task) == 'pending':
+        progress = session.xenapi.task.get_progress(task)
+        time.sleep(1)
+    return show_instance(name)
+
+def suspend(name,call=None,session=None):
+    '''
+    Suspend a vm to disk
+
+    .. code-block:: bash
+
+        salt-cloud -a suspend xenvm01
+
+    '''
+    if call == 'function':
+        raise SaltCloudException(
+                'The show_instnce function must be called with -a or --action.'
+        )
+    if session is None:
+        session = _get_session()
+    log.info('Suspending VM {0}'.format(name))
+    vm = _get_vm(name,session)
+    task = session.xenapi.Async.VM.suspend(vm)
+    while session.xenapi.task.get_status(task) == 'pending':
+        progress = session.xenapi.task.get_progress(task)
+        time.sleep(1)
+    return show_instance(name)
+
+def resume(name,call=None,session=None):
+    '''
+    Resume a vm from disk
+
+    .. code-block:: bash
+
+        salt-cloud -a resume xenvm01
+
+    '''
+    if call == 'function':
+        raise SaltCloudException(
+                'The show_instnce function must be called with -a or --action.'
+        )
+    if session is None:
+        session = _get_session()
+    log.info('Resuming VM {0}'.format(name))
+    vm = _get_vm(name,session)
+    task = session.xenapi.Async.VM.resume(vm,False,True)
+    while session.xenapi.task.get_status(task) == 'pending':
```

Comment:
There seems to be a lot of duplication of this loop. Could we just make a helper func to do it?